### PR TITLE
Add proxyless endpoint support

### LINKED ISSUE #1637: Support proxy-less endpoints
This issue outlines how networking works in aspire https://github.com/dotnet/docs-aspire/issues/232. Today we spin up a proxy for each endpoint which means the application being proxied needs to be amenable to accepting a port to listen on. The proxy is important for 2 reasons:
1. We can avoid forwarding the connection until the underling endpoint is ready
2. We can route to multiple replicas of the underlying service

The scenario that does not work is the one described here https://github.com/dotnet/docs-aspire/issues/232#issuecomment-1884575748

> We have also a legacy java http service running on specific port on Windows.
We add the service as a console project managing its execution.
How can I specify its port in WithServiceBinding if it is not running in a container?
If i set the hostPort to the service port, the proxy blocks the port and the service cannot be started.

In this scenario, there can be no proxy because the underlying process can't listen to anything but the port it already has. The only work around for this today is to avoid using WithEndpoint (formerly WithServiceBinding) but that breaks the ability to add references to other resources (WithReference) and it also means the dashboard doesn't show the url for the service.

Open questions:

- What happens when a resource has endpoints that are proxy-less and proxied? This probably shouldn't be allowed.

### PR DESCRIPTION
**Pull request by BrennanConroy**
Closes https://github.com/dotnet/aspire/issues/1637
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2314)

### CODE REVIEW

**Comment by davidfowl** [comment]
Can you make a playground example with this feature?

File: playground/ProxylessEndToEnd/ProxylessEndToEnd.AppHost/Program.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+var builder = DistributedApplication.CreateBuilder(args);
+
+builder.AddProject<Projects.ProxylessEndToEnd_ApiService>("api")
+    .WithEndpoint("http", ea =>
+    {
+        ea.UriScheme = "http";
+        ea.Port = 12345;
+        ea.IsProxied = false;
```
Comment (davidfowl) [inline playground/ProxylessEndToEnd/ProxylessEndToEnd.AppHost/Program.cs:11]: Is this the only way to use it? Can you set it using the other WithEndpoint overloads?

File: playground/ProxylessEndToEnd/ProxylessEndToEnd.AppHost/Program.cs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+var builder = DistributedApplication.CreateBuilder(args);
+
+builder.AddProject<Projects.ProxylessEndToEnd_ApiService>("api")
+    .WithEndpoint("http", ea =>
+    {
+        ea.UriScheme = "http";
```
Comment (davidfowl) [inline playground/ProxylessEndToEnd/ProxylessEndToEnd.AppHost/Program.cs:9]: Is this required?

File: src/Aspire.Hosting/Dcp/ApplicationExecutor.cs [hunk 3]
```diff
@@ -7,7 +7,6 @@
 using Aspire.Hosting.Lifecycle;
 using Aspire.Hosting.Utils;
 using k8s;
-//using k8s.Models;
```
Comment (davidfowl) [inline src/Aspire.Hosting/Dcp/ApplicationExecutor.cs]: ?

**Comment by davidfowl** [comment]
A couple of questions:

1. What happens to the service discovery information with this change? Does it work well? Does it affect anything?
2. What happens if I set IsProxied false and don't choose a host port? What happens when this is a container?
3. Do we still get an AllocatedEndpoint annotation when IsProxied = false?

File: playground/ProxylessEndToEnd/ProxylessEndToEnd.AppHost/Program.cs (hunk 2 above)
Comment (BrennanConroy) [reply playground/ProxylessEndToEnd/ProxylessEndToEnd.AppHost/Program.cs:9]: Yes. We do "magic" today in the other WithEndpoint overload where the name is also the scheme.

File: playground/ProxylessEndToEnd/ProxylessEndToEnd.AppHost/Program.cs (hunk 1 above)
Comment (davidfowl) [reply playground/ProxylessEndToEnd/ProxylessEndToEnd.AppHost/Program.cs:11]: Can we add more the optional parameter to the base overload?

**Comment by BrennanConroy** [comment]
> What happens to the service discovery information with this change? Does it work well? Does it affect anything?

Since I'm not an expert here I can't say for sure, but referencing a proxyless endpoint from another project works so far.

> * What happens if I set IsProxied false and don't choose a host port? What happens when this is a container?

Throws for missing port. Containers aren't proxied so the flag does nothing I think? Will have to verify behavior here.

> * Do we still get an AllocatedEndpoint annotation when IsProxied = false?

Yes, that's needed for `.WithReference(projectWithProxylessEndpoint)`

**Comment by davidfowl** [comment]
> Since I'm not an expert here I can't say for sure, but referencing a proxyless endpoint from another project works so far.

This is what matters. We just need to make sure that calling WithReference continues to work and that URLs show up in the dashboard.

> Throws for missing port. Containers aren't proxied so the flag does nothing I think? Will have to verify behavior here.

Containers are proxied by default. We have 3 ports:
1. The proxy port
2. The host port (you have to look in docker desktop to see this today)
3. The container port

My assumption is that when you say IsProxied=false, for a container, should throw and force you to specify a host port. 

> Yes, that's needed for .WithReference(projectWithProxylessEndpoint)

Excellent. Can you add a container with a proxyless port in your playground sample.

**Review by karolz-ms — APPROVED** [review]
ApplicationExecutor changes look good

File: tests/Aspire.Hosting.Tests/DistributedApplicationTests.cs [hunk 4]
```diff
@@ -454,6 +455,185 @@ await foreach(var resource in s.WatchAsync<Executable>(cancellationToken: token)
         }
     }
 
+    [LocalOnlyFact("docker")]
```
Comment (davidfowl) [inline tests/Aspire.Hosting.Tests/DistributedApplicationTests.cs:458]: Do you have any tests that use the new APIs?

File: tests/Aspire.Hosting.Tests/DistributedApplicationTests.cs [hunk 5]
```diff
@@ -483,6 +484,185 @@ await foreach(var resource in s.WatchAsync<Executable>(cancellationToken: token)
...
+        Assert.Equal("localhost:1234", env.Value);
+
+        var list = await s.ListAsync<Container>();
+        var redisContainer = Assert.Single(list);
+        Assert.Equal(1234, Assert.Single(redisContainer.Spec.Ports!).HostPort);
+
+        await app.StopAsync();
+    }
+
+    [Fact]
```
Comment (BrennanConroy) [inline tests/Aspire.Hosting.Tests/DistributedApplicationTests.cs:649]:
```suggestion
    [LocalOnlyFact("docker")]
```

**Review by davidfowl — APPROVED** [review]

**Pull request merged**